internal/eventstream/client.go

```diff
@@ -0,0 +1,326 @@
+
+		eventTags = map[string]string{
+			"host":    host,
+			"service": service,
+		}
+
+		serviceGroups, err := client.fetchServiceGroups(host, service)
+		if err != nil {
+			return nil, err
+		}
+		for _, serviceGroup := range serviceGroups {
+			eventExtraTags["servicegroup/"+serviceGroup] = ""
+		}
+	} else {
+		eventName = host
+		eventUrlSuffix = "/icingadb/host?name=" + url.PathEscape(host)
+
+		eventTags = map[string]string{
+			"host": host,
+		}
+	}
+
+	hostGroups, err := client.fetchHostGroups(host)
+	if err != nil {
+		return nil, err
+	}
+	for _, hostGroup := range hostGroups {
+		eventExtraTags["hostgroup/"+hostGroup] = ""
+	}
+
+	return &event.Event{
+		Time:      time.Now(),
+		SourceId:  client.IcingaNotificationsEventSourceId,
+		Name:      eventName,
+		URL:       client.IcingaWebRoot + eventUrlSuffix,
+		Tags:      eventTags,
+		ExtraTags: eventExtraTags,
+	}, nil
+}
+
+// buildHostServiceEvent constructs an event.Event based on a CheckResult, a Host or Service state, a Host name and an
+// optional Service name if the Event should represent a Service object.
+func (client *Client) buildHostServiceEvent(result CheckResult, state int, host, service string) (*event.Event, error) {
+	var eventSeverity event.Severity
+
+	if service != "" {
+		switch state {
+		case 0:
+			eventSeverity = event.SeverityOK
+		case 1:
+			eventSeverity = event.SeverityWarning
+		case 2:
+			eventSeverity = event.SeverityCrit
+		default:
+			eventSeverity = event.SeverityErr
+		}
+	} else {
+		switch state {
+		case 0:
+			eventSeverity = event.SeverityOK
+		case 1:
+			eventSeverity = event.SeverityCrit
+		default:
+			eventSeverity = event.SeverityErr
+		}
+	}
+
+	ev, err := client.buildCommonEvent(host, service)
+	if err != nil {
+		return nil, err
+	}
+
+	ev.Type = event.TypeState
+	ev.Severity = eventSeverity
+	ev.Message = result.Output
+
+	return ev, nil
+}
+
+// buildAcknowledgementEvent from the given fields.
+func (client *Client) buildAcknowledgementEvent(host, service, author, comment string) (*event.Event, error) {
+	ev, err := client.buildCommonEvent(host, service)
+	if err != nil {
+		return nil, err
+	}
+
+	ev.Type = event.TypeAcknowledgement
+	ev.Username = author
+	ev.Message = comment
+
+	return ev, nil
+}
+
+// eventDispatcher receives generated event.Events to be either buffered or directly delivered to the CallbackFn.
+//
+// When the Client is in the replay phase, events from the Event Stream API will be cached until the replay phase has
+// finished, while replayed events will be delivered directly.
+func (client *Client) eventDispatcher() {
+	var replayBuffer []*event.Event
+
+	for {
+		select {
+		case <-client.Ctx.Done():
+			client.Logger.Warnw("Closing event dispatcher as context is done", zap.Error(client.Ctx.Err()))
+			return
+
+		case <-client.replayTrigger:
+			for _, ev := range replayBuffer {
+				client.CallbackFn(ev)
+			}
```

Comment:
As just discussed, these should not just all be replayed, as there could be events that are duplicates, or worse, states that are older than what was read from the objects API. So there should probably be some `map[object]timestamp` to remove them.